Add doc for using GPU in Knative (#741)
vhive-serverless/vHive: docs/developers_guide.md

Follow the guide to [setup stock knative](#testing-stock-knative-setup-or-images). 

``` bash
./scripts/cloudlab/setup_node.sh stock-only
```

### Install Nvidia Container Toolkit and Modify Containerd Settings

You can use the script provided if the install of containerd is using  our script or manually edit the containerd settings following [Nvidia's official document](https://docs.nvidia.com/datacenter/cloud-native/kubernetes/install-k8s.html#install-nvidia-container-toolkit-nvidia-docker2).

``` bash
./scripts/gpu/setup_nvidiacontainerd.sh
```

### Start Containerd and Knative

``` bash
sudo screen -dmS containerd containerd; sleep 5;
./scripts/cluster/create_one_node_cluster.sh stock-only
```

### Install NVIDIA Device Plugin

Install the Nvidia Device plugin after all pods are running or completed.

``` bash
./scripts/gpu/setup_nvidiadevice.sh
```
At this point, all pods should be successfully deployed.
``` bash
$ kubectl get pods -A
NAMESPACE              NAME                                       READY   STATUS      RESTARTS   AGE
nvidia-device-plugin   nvidia-device-plugin-1684892866-rtgk9      1/1     Running     0          80s

```

> And you can use the gpu-pod.yaml to test whether the Nvidia Device Plugin works or not

Reviewer: ```suggestion
And you can use the gpu-pod.yaml to test whether the Nvidia Device Plugin works
```

``` bash
kubectl apply -f ./scripts/gpu/gpu-pod.yaml
```

> After it completed, check the log and it should be

Reviewer: ```suggestion
After that, check the log to contain:
```

``` bash
$ kubectl logs gpu-operator-test
```
``` bash
[Vector addition of 50000 elements]
Copy input data from the host memory to the CUDA device
CUDA kernel launch with 196 blocks of 256 threads
Copy output data from the CUDA device to the host memory
Test PASSED
Done
```
### Test Knative with GPU

> Using GPU in Knative is simple and same to the method in Kubernetes. The only change is to add the limits of GPU under the containers part.

Reviewer: ```suggestion
Using GPU in Knative is simple and similar as for a regular Kubernetes service. The only change is to add the limits of GPU to the YAML file.
```

``` yaml
resources:
  limits:
    nvidia.com/gpu: 1
```

> You can also try deploying our container, which is a simple Go service that calls and returns the output of nvidia-smi.

Reviewer: ```suggestion
You can also deploy our example container, which is a Golang function that calls and returns the output of nvidia-smi.
```

``` bash

> kn service apply -f ./scripts/gpu/gpu-service.yaml

Reviewer: is it a service or a Knative function? we need only the function example

Author: It's a service. For Knative function version, it needs extra dependency for kn func. I'll try to change it to a function example

| Fan  Temp  Perf            Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  Tesla P100-PCIE-16GB            On | 00000000:00:06.0 Off |                  Off |
| N/A   35C    P0               27W / 250W|      4MiB / 16384MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|

> +---------------------------------------------------------------------------------------+

Reviewer: enough to show the header of the output, no need to show a full output.

to speed up testing environment setup requiring Kubernetes.

* Current [eStargz](https://github.com/containerd/stargz-snapshotter) version is 0.13.0.

## Deploying Knative functions that require GPU Support

Knative functions can use GPU although only `stock-only` mode is supported. 

### Install Stock Kubernetes and Knative

Follow the guide to [setup stock knative](#testing-stock-knative-setup-or-images). 

``` bash
./scripts/cloudlab/setup_node.sh stock-only
```

### Install NVIDIA Driver and NVIDIA Container Toolkit 

> The script will install NVIDIA CUDA Driver and assume there’s no NVIDIA driver currently running. You can use the script provided if the install of containerd is using our script or manually edit the containerd settings following [NVIDIA's official document](https://docs.nvidia.com/datacenter/cloud-native/kubernetes/install-k8s.html#install-nvidia-container-toolkit-nvidia-docker2).

Reviewer: please specify the specs of the platform, i.e., CUDA and GPU version, on which you tested the following commands. Give recommendations on which platforms/GPUs the commands should work. 

split the long line into several lines for finer-grain version control

``` bash
$ kubectl get pods -A
NAMESPACE              NAME                                       READY   STATUS      RESTARTS   AGE
nvidia-device-plugin   nvidia-device-plugin-1684892866-rtgk9      1/1     Running     0          80s
```
And you can use the gpu-pod.yaml to test whether the NVIDIA Device Plugin works

``` bash
kubectl apply -f ./configs/gpu/gpu-pod.yaml
```
After that, check the log to contain:

``` bash
$ kubectl logs gpu-operator-test
```
``` bash
[Vector addition of 50000 elements]
Copy input data from the host memory to the CUDA device
CUDA kernel launch with 196 blocks of 256 threads
Copy output data from the CUDA device to the host memory
Test PASSED
Done
```
### Test Knative with GPU
Using GPU in Knative is simple and similar as for a regular Kubernetes service. The only change is to add the limits of GPU to the YAML file.

``` yaml
resources:
  limits:
    nvidia.com/gpu: 1
```
You can also deploy our example container, which is a Golang function that calls and returns the output of `nvidia-smi`.

``` bash
kn service apply -f ./configs/gpu/gpu-function.yaml
```
Once the service has been deployed successfully, you can call it and check it's response.
``` bash

> curl "$(kn service describe hello-gpu -o url)"

Reviewer: ```suggestion
curl "$(kn service describe hello-gpu -o <URL>)"
```
